scala/src/com/google/idea/blaze/scala/run/producers/DeployableJarRunConfigurationProducer.java

```diff
@@ -95,46 +90,34 @@ protected boolean doIsConfigFromContext(
     if (mainObject == null) {
       return false;
     }
-    TargetIdeInfo target = findTarget(context.getProject(), mainObject);
+    TargetInfo target = findTarget(context.getProject(), mainObject);
     if (target == null) {
       return false;
     }
-    Label label = target.key.label;
-    File jarFile = getDeployJarFile(label, context.getProject());
-    if (jarFile == null) {
-      return false;
-    }
+
     return mainClass.getQualifiedName().equals(mainObject.getTruncedQualifiedName())
-        && configuration.getVMParameters().contains("-cp " + jarFile.getPath());
+        && configuration.getVMParameters().endsWith(
+                String.format("%s_deploy.jar", target.label.targetName()));
   }
 
   @Nullable
-  private static TargetIdeInfo findTarget(Project project, ScObject mainObject) {
+  private static TargetInfo findTarget(Project project, ScObject mainObject) {
     File mainObjectFile = RunUtil.getFileForClass(mainObject);
     if (mainObjectFile == null) {
       return null;
     }
 
-    Collection<TargetIdeInfo> targets =
-        BlazeScalaMainClassRunConfigurationProducer.findScalaBinaryTargets(project, mainObjectFile);
+    Collection<TargetInfo> targets =
+            SourceToTargetFinder.findTargetsForSourceFile(
```

Comment:
This is going to find all library rules reachable from the current source file (in the transitive closure). Is that actually what you want here?

SourceToTargetProvider (the almost duplicate APIs are confusing and need to be redone) will return only targets directly building the given source file -- is that enough?

Reply:
Hmm, I need to check, thanks for the heads up. Will report back shortly.